src/github_sdk.py

```diff
@@ -1,101 +1,135 @@
+            self.sentry_project_url = f"{base_uri}/api/{project_id}/envelope/"
+
+    def _fetch_github(self, url):
+        headers = {}
+        if self.token:
+            headers["Authorization"] = f"token {self.token}"
+        req = requests.get(url, headers=headers)
+        req.raise_for_status()
+        return req
+
+    def _get_extra_metadata(self, job):
+        # XXX: This is the slowest call
+        runs = self._fetch_github(job["run_url"]).json()
+        workflow = self._fetch_github(runs["workflow_url"]).json()
+        repo = runs["repository"]["full_name"]
+        meta = {
+            # "workflow_name": workflow["name"],
+            "author": runs["head_commit"]["author"],
+            # https://getsentry.atlassian.net/browse/TET-22
+            # Tags are not linkified externally, plain text data can be selected in browsers and opened
+            "data": {
+                "job": job["html_url"],
+            },
+            "tags": {
+                "job_status": job["conclusion"],  # e.g. success, failure, skipped
+                "branch": runs["head_branch"],
+                "commit": runs["head_sha"],
+                "repo": repo,
+                "run_attempt": runs["run_attempt"],  # Rerunning a job
+                # It allows querying jobs within the same workflow (e.g. foo.yml)
+                "workflow": workflow["path"].rsplit("/")[-1],
+            },
+        }
+        if runs.get("pull_requests"):
+            pr_number = runs["pull_requests"][0]["number"]
+            meta["data"]["pr"] = f"https://github.com/{repo}/pull/{pr_number}"
+            meta["tags"]["pull_request"] = pr_number
+
+        return meta
+
+    # Documentation about traces, transactions and spans
+    # https://docs.sentry.io/product/sentry-basics/tracing/distributed-tracing/#traces
+    # https://develop.sentry.dev/sdk/performance/
+    def _generate_trace(self, job):
+        meta = self._get_extra_metadata(job)
+        transaction = _base_transaction(job)
+        transaction["user"] = meta["author"]
+        transaction["tags"] = meta["tags"]
+        transaction["contexts"]["trace"]["data"] = meta["data"]
+
+        # Transactions have name, spans don't.
+        transaction["contexts"]["trace"]["op"] = job["name"]
+        transaction["contexts"]["trace"]["description"] = job["name"]
+        transaction["contexts"]["trace"][
+            "status"
+        ] = self.github_status_trace_status.get(job["conclusion"], "unimplemented")
+        transaction["spans"] = _generate_spans(
+            job["steps"],
+            transaction["contexts"]["trace"]["span_id"],
+            transaction["contexts"]["trace"]["trace_id"],
+        )
+        return transaction
+
+    def _send_envelope(self, trace):
+        envelope = Envelope()
+        envelope.add_transaction(trace)
+        now = datetime.utcnow()
+
+        headers = {
+            "event_id": get_uuid(),  # Does this have to match anything?
```

Comment:
You will notice in the fixture that I add one more `uuid` value since we now test this code.